fix($location): don't crash if navigating outside the app base

petebacondarwin opened the pull request:
This should fix #11667

gkalpak commented on test/ng/locationSpec.js:
```diff
@@ -601,35 +601,35 @@ describe('$location', function() {
 
       it('should return decoded characters for search specified in URL', function() {
-        var locationUrl = new LocationHtml5Url('http://host.com/');
+        var locationUrl = new LocationHtml5Url('http://host.com/', 'http://host.com/', 'http://host.com/');
```
Is the `basePrefix` really needed here ?

petebacondarwin replied:
Oops. No that was a bit of over zealous cutting and pasting :-)

petebacondarwin committed:
fix($location): don't crash if navigating outside the app base

Previously, if you navigate outside of the Angular application, say be clicking
the back button, the $location service would try to handle the url change
and error due to the URL not being valid for the application.

This fixes that issue by ensuring that a reload happens when you navigate
to a URL that is not within the application.

Closes #11667

petebacondarwin committed:
refactor($location): compute `appBaseNoFile` only once

petebacondarwin committed:
refactor($location): compute `appBaseNoFile` only once

Narretz commented:
Is this resolved by https://github.com/angular/angular.js/commit/9e492c358c19549696577c86c2c61b93f50ab356 and https://github.com/angular/angular.js/commit/92c7ce5becb9b90216c49f50e0f57f5dbeda8eba?

petebacondarwin commented:
Yes.
